ndnsd/discovery/service-discovery.cpp

```diff
@@ -0,0 +1,374 @@
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+ * PURPOSE.  See the GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along with
+ * NDNSD, e.g., in COPYING.md file.  If not, see <http://www.gnu.org/licenses/>.
+ **/
+
+#include "service-discovery.hpp"
+#include <string>
+#include <iostream>
+
+using namespace ndn::time_literals;
+
+namespace ndnsd {
+namespace discovery {
+
+// consumer
+ServiceDiscovery::ServiceDiscovery(const ndn::Name& serviceName,
+                                   const std::map<char, uint8_t>& pFlags,
+                                   const DiscoveryCallback& discoveryCallback)
+  : m_serviceName(serviceName)
+  , m_appType(processFalgs(pFlags, 't'))
+  , m_counter(0)
+  , m_syncProtocol(processFalgs(pFlags, 'p'))
+  , m_syncAdapter(m_face, m_syncProtocol, makeSyncPrefix(m_serviceName),
+                  "/defaultName", 1600_ms,
+                  std::bind(&ServiceDiscovery::processSyncUpdate, this, _1))
+  , m_discoveryCallback(discoveryCallback)
+{
+}
+
+// producer
+ServiceDiscovery::ServiceDiscovery(const std::string& filename,
+                                   const std::map<char, uint8_t>& pFlags,
+                                   const DiscoveryCallback& discoveryCallback)
+  : m_filename(filename)
+  , m_fileProcessor(m_filename)
+  , m_appType(processFalgs(pFlags, 't'))
+  , m_syncProtocol(processFalgs(pFlags, 'p'))
+  , m_syncAdapter(m_face, m_syncProtocol, makeSyncPrefix(m_fileProcessor.getServiceName()),
+                  m_fileProcessor.getAppPrefix(), 1600_ms,
+                  std::bind(&ServiceDiscovery::processSyncUpdate, this, _1))
+  , m_discoveryCallback(discoveryCallback)
+{
+  setUpdateProducerState();
+  setInterestFilter(m_producerState.applicationPrefix);
+
+  // listen on reload prefix as well.
+  setInterestFilter(ndnsd::discovery::NDNSD_RELOAD_PREFIX);
+}
+
+void
+ServiceDiscovery::setUpdateProducerState(bool update)
+{
+  NDNSD_LOG_INFO("Setting/Updating producers state: ");
+  if (update)
+  {
+    m_fileProcessor.processFile();
+  }
+  m_producerState.serviceName = m_fileProcessor.getServiceName();
+  m_producerState.applicationPrefix = m_fileProcessor.getAppPrefix();
+  m_producerState.serviceLifetime = m_fileProcessor.getServiceLifetime();
+  m_producerState.publishTimestamp = ndn::time::system_clock::now();
+  m_producerState.serviceMetaInfo = m_fileProcessor.getServiceMeta();
+}
+
+ndn::Name
+ServiceDiscovery::makeSyncPrefix(ndn::Name& service)
+{
+  ndn::Name sync("/discovery");
+  sync.append(service);
+  return sync;
+}
+
+uint8_t
+ServiceDiscovery::processFalgs(const std::map<char, uint8_t>& flags, const char type)
+{
+  auto key = flags.find(type);
+  if (key != flags.end())
+  {
+    return flags.find(type)->second;
+  }
+  else
+  {
+    NDN_THROW(Error("Flag type not found!"));
+    NDNSD_LOG_ERROR("Flag type not found!");
```

Comment:
Unreachable.

Reply:
won't it reach if the specific key is not found? it should right?